taichi/backends/metal/codegen_metal.cpp

```diff
@@ -551,6 +552,65 @@ class KernelCodegen : public IRVisitor {
     TI_WARN("Cannot print inside Metal kernel, ignored");
   }
 
+  void visit(StackAllocaStmt *stmt) override {
+    TI_ASSERT(stmt->width() == 1);
+
+    const auto &var_name = stmt->raw_name();
+    emit("byte {}[{}];", var_name, stmt->size_in_bytes());
+    emit("mtl_ad_stack_init({});", var_name);
+  }
+
+  void visit(StackPopStmt *stmt) override {
+    emit("mtl_ad_stack_pop({});", stmt->stack->raw_name());
+  }
+
+  void visit(StackPushStmt *stmt) override {
+    auto *stack = stmt->stack->as<StackAllocaStmt>();
+    const auto &stack_name = stack->raw_name();
+    const auto elem_size = stack->element_size_in_bytes();
+    emit("mtl_ad_stack_push({}, {});", stack_name, elem_size);
```

Comment:
If I'm thinking right, elem_size will always be 4 in Metal since it only support 32-bit data?

Reply:
Yeah, I'd say it's *almost* always, but it doesn't prevent us from using more complicated types (For example, one day we may support composite types like `complex {f32, f32}`?) 
Metal only supports 32-bit **primitive** data types, but it's fine to have composite types.